File: swift/Tests/JsonTagTests/SerializeTests.swift
import XCTest
import JsonTag

final class SerializeTests: XCTestCase {

    private let encoder = JsonTag.makeEncoder()

    // Interop JSON — identical strings used across all 12 language libraries
    let circleJson = """
    {"#type":"Circle","radius":4}
    """
    let rectangleJson = """
    {"#type":"Rectangle","width":10,"height":5}
    """
    let carJson = """
    {"#type":"Car","make":"Toyota"}
    """
    let truckJson = """
    {"#type":"Truck","payloadTons":5}
    """

    // JSONEncoder with .sortedKeys sorts fields alphabetically: height before width
    let swiftRectangleJson = """
    {"#type":"Rectangle","height":5,"width":10}
    """

    func testSerialize_whenSerializingCircle_matchesInteropFormat() throws {
        // Act
        let json = String(data: try encoder.encode(Shape.circle(radius: 4)), encoding: .utf8)!

        // Assert
        XCTAssertEqual(json, circleJson)
    }

    func testSerialize_whenSerializingRectangle_matchesInteropFormat() throws {
        // Act
        let json = String(data: try encoder.encode(Shape.rectangle(width: 10, height: 5)), encoding: .utf8)!

        // Assert
        XCTAssertEqual(json, swiftRectangleJson)
    }

    func testSerialize_whenSerializingCar_matchesInteropFormat() throws {
        // Act
        let json = String(data: try encoder.encode(Vehicle.personalVehicle(make: "Toyota")), encoding: .utf8)!

        // Assert
        XCTAssertEqual(json, carJson)
    }

    func testSerialize_whenSerializingTruck_matchesInteropFormat() throws {
        // Act
        let json = String(data: try encoder.encode(Vehicle.transportationVehicle(payloadTons: 5)), encoding: .utf8)!

        // Assert
        XCTAssertEqual(json, truckJson)
    }
}

File: rust/json-tag-tests/tests/serialize_tests.rs
mod test_types;

use test_types::{Shape, Vehicle};

// Interop JSON — identical strings used across all 12 language libraries
const CIRCLE_JSON: &str = r##"{"#type":"Circle","radius":4}"##;
const RECTANGLE_JSON: &str = r##"{"#type":"Rectangle","width":10,"height":5}"##;
const CAR_JSON: &str = r##"{"#type":"Car","make":"Toyota"}"##;
const TRUCK_JSON: &str = r##"{"#type":"Truck","payloadTons":5}"##;

// serde_json emits .0 for whole-number f64 values
const SERDE_CIRCLE_JSON: &str = r##"{"#type":"Circle","radius":4.0}"##;
const SERDE_RECTANGLE_JSON: &str = r##"{"#type":"Rectangle","width":10.0,"height":5.0}"##;
const SERDE_TRUCK_JSON: &str = r##"{"#type":"Truck","payloadTons":5.0}"##;

#[test]
fn serialize_when_serializing_circle_matches_interop_format() {
    // Act
    let json = serde_json::to_string(&Shape::Circle { radius: 4.0 }).unwrap();

    // Assert
    assert_eq!(json, SERDE_CIRCLE_JSON);
}

#[test]
fn serialize_when_serializing_rectangle_matches_interop_format() {
    // Act
    let json = serde_json::to_string(&Shape::Rectangle {
        width: 10.0,
        height: 5.0,
    })
    .unwrap();

    // Assert
    assert_eq!(json, SERDE_RECTANGLE_JSON);
}

#[test]
fn serialize_when_serializing_car_matches_interop_format() {
    // Act
    let json = serde_json::to_string(&Vehicle::PersonalVehicle {
        make: "Toyota".to_string(),
    })
    .unwrap();

    // Assert
    assert_eq!(json, CAR_JSON);
}

#[test]
fn serialize_when_serializing_truck_matches_interop_format() {
    // Act
    let json = serde_json::to_string(&Vehicle::TransportationVehicle {
        payload_tons: 5.0,
    })
    .unwrap();

    // Assert
    assert_eq!(json, SERDE_TRUCK_JSON);
}
